logictest: add 3node-tenant as default config

asubiotto opened the pull request:
Closes #47901

The logic test directive now supports syntax to blacklist a configuration, i.e.
running all specified configurations apart from a blacklisted configuration:
```
 # LogicTest: default-configs !3node-tenant
```
If a blacklist is specified without an accompanying configuration, the
default config is assumed. i.e., the following directive is equivalent to the
one above:
```
 # LogicTest: !3node-tenant
```

The second commit adds the `3node-tenant` config to the list of default
configs and adds it as a blacklist in any files that fail under this config.

Release note: None (testing change)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48530)
<!-- Reviewable:end -->

tbg reviewed (approved):
Looks good! What are the next steps? I assume you'll look into the cluster settings thing. After that, are you planning to revisit the skipped files here and to isolate the failures more?

> Reviewed 1 of 1 files at r1, 28 of 28 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48530#-:-M7D032Y0ngfNH98iWF3:b-ecm77s)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @nvanbenschoten)

___
*[pkg/sql/logictest/logic.go, line 570 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48530#-M7D-xAH94deZ4CAhVcl:-M7D-xAH94deZ4CAhVcm:b1g1wae) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb7f0c2b3f64cb1be4edcb5ef08ffeb27989a03e/pkg/sql/logictest/logic.go#L570)):*
> ```Go
> 		name:              "3node-tenant",
> 		numNodes:          3,
> 		overrideAutoStats: "false",
> ```

This gets more tests passing by default despite the SET CLUSTER SETTING not working? Add a comment.
___
*[pkg/sql/logictest/logic.go, line 1417 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48530#-M7CzVBaCKg6qx9OiKP3:-M7CzVBaCKg6qx9OiKP4:bsa0jyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb7f0c2b3f64cb1be4edcb5ef08ffeb27989a03e/pkg/sql/logictest/logic.go#L1417)):*
> ```Go
> 		if !ok {
> 			switch configName {
> 			case defaultConfigName:
> ```

So `defaultConfigName` is not an existing config? Ah, I see, it's too magical. Carry on.
___
*[pkg/sql/logictest/testdata/logic_test/cascade, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48530#-M7D-AC_86JkCKAIUkKB:-M7D-AC_86JkCKAIUkKC:bzs0ne) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3364cebdcb1fb0fb923babb3dc6367234c157fd/pkg/sql/logictest/testdata/logic_test/cascade#L1)):*
> ```
> # This test times out when run with 3node-tenant.
> ```

Are you planning to revisit all of the ones for which the failure is unknown? Either way it might be worth filing an issue for this under the A-multitenancy{,-blocker} label.


<!-- Sent from Reviewable.io -->

tbg commented:
Oh - and another question: if we run the logic tests with `--config 3node-tenant` will it also run the ones with the `!3node-tenant` header? Not sure what behavior we want here. I think we'd still want to obey the header, right? Just checking that this is now what we're getting.

asubiotto reviewed (commented):
I was wondering about the next steps. Triaging seems to be the most important thing to do next, although it is going to be a pain because https://github.com/cockroachdb/cockroach/issues/47892 seems to be a cause of failure for a lot of tests (the node ID change looks like it traded errors for timeouts, reverting that change and rerunning to confirm) so it might be more productive to hold off and work on fixes for two failure causes I understand and can fix: cluster settings and EXPLAIN using a node ID.

Yes, specifying `--config` overrides the directive on whatever files you're running, which I think is the reason the flag was introduced.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48530#-:-M7DjyNI83iK2h3sawvO:b5pu8ef)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/sql/logictest/logic.go, line 570 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48530#-M7D-xAH94deZ4CAhVcl:-M7DhgXB9Jur0fv8ILnd:b2qv4l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb7f0c2b3f64cb1be4edcb5ef08ffeb27989a03e/pkg/sql/logictest/logic.go#L570)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This gets more tests passing by default despite the SET CLUSTER SETTING not working? Add a comment.
</blockquote></details>

Yes, so I was sort of confused when I mentioned this during the meeting. This knob sets the cluster setting on the nodes acting as kv servers, which works and fixes the flakiness which caused the introduction of this knob in all configs. The thing that does not work is running `SET CLUSTER SETTING` within a test, which runs that on the tenant, as well as things like `SHOW JOBS`, since that returns some auto stats jobs.
___
*[pkg/sql/logictest/logic.go, line 1417 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48530#-M7CzVBaCKg6qx9OiKP3:-M7DihRu-vRQlAY2PGPV:baimlzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb7f0c2b3f64cb1be4edcb5ef08ffeb27989a03e/pkg/sql/logictest/logic.go#L1417)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

So `defaultConfigName` is not an existing config? Ah, I see, it's too magical. Carry on.
</blockquote></details>

This and `fiveNodeDefaultConfigName` are syntactic sugar to specify all the default configs to allow for composition.
___
*[pkg/sql/logictest/testdata/logic_test/cascade, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48530#-M7D-AC_86JkCKAIUkKB:-M7DiySR-523Vn9-xV6L:b6rfyt4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3364cebdcb1fb0fb923babb3dc6367234c157fd/pkg/sql/logictest/testdata/logic_test/cascade#L1)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Are you planning to revisit all of the ones for which the failure is unknown? Either way it might be worth filing an issue for this under the A-multitenancy{,-blocker} label.
</blockquote></details>

Yes, the goal of this PR is just to add `3node-tenant` as the default config with triaging to be done later.


<!-- Sent from Reviewable.io -->

tbg commented:
Yeah let's hold off until https://github.com/cockroachdb/cockroach/pull/48795 is in. I think that might "fix" jobs! Let's talk more about next steps tomorrow but looking into the cluster settings work is definitely appropriate.

Commit:
logictest: add 3node-tenant as default config

This commit also blacklists 3node-tenant in certain logic test files that would
fail otherwise.

Release note: None

Commit:
logictest: add option to blacklist configs to logictest directive

The logic test directive now supports syntax to blacklist a configuration, i.e.
running all specified configurations apart from a blacklisted configuration:
 # LogicTest: default-configs !3node-tenant
If a blacklist is specified without an accompanying configuration, the
default config is assumed. i.e., the following directive is equivalent to the
one above:
 # LogicTest: !3node-tenant

Release note: None (testing change)

asubiotto commented:
TFTR
bors r=tbg
